src/Umbraco.Infrastructure/Compose/DatabaseServerRegistrarAndMessengerComponent.cs

```diff
@@ -81,201 +75,48 @@ public override void Compose(Composition composition)
-        private readonly BackgroundTaskRunner<IBackgroundTask> _processTaskRunner;
-        private bool _started;
-        private IBackgroundTask[] _tasks;
         private readonly IRequestAccessor _requestAccessor;
 
         public DatabaseServerRegistrarAndMessengerComponent(
-            IServerRegistrar serverRegistrar,
             IServerMessenger serverMessenger,
-            IServerRegistrationService registrationService,
-            ILogger<DatabaseServerRegistrarAndMessengerComponent> logger,
-            ILoggerFactory loggerFactory,
-            IApplicationShutdownRegistry hostingEnvironment,
             IRequestAccessor requestAccessor)
         {
-            _logger = logger;
-            _loggerFactory = loggerFactory;
-            _registrationService = registrationService;
             _requestAccessor = requestAccessor;
-
-            // create task runner for DatabaseServerRegistrar
-            _registrar = serverRegistrar as DatabaseServerRegistrar;
-            if (_registrar != null)
-            {
-                _touchTaskRunner = new BackgroundTaskRunner<IBackgroundTask>("ServerRegistration",
-                    new BackgroundTaskRunnerOptions { AutoStart = true }, _loggerFactory.CreateLogger<BackgroundTaskRunner<IBackgroundTask>>(), hostingEnvironment);
-            }
-
-            // create task runner for BatchedDatabaseServerMessenger
             _messenger = serverMessenger as IBatchedDatabaseServerMessenger;
-            if (_messenger != null)
-            {
-                _processTaskRunner = new BackgroundTaskRunner<IBackgroundTask>("ServerInstProcess",
-                    new BackgroundTaskRunnerOptions { AutoStart = true }, _loggerFactory.CreateLogger<BackgroundTaskRunner<IBackgroundTask>>(), hostingEnvironment);
-            }
         }
 
         public void Initialize()
         {
-            //We will start the whole process when a successful request is made
-            if (_registrar != null || _messenger != null)
-                _requestAccessor.RouteAttempt += RegisterBackgroundTasksOnce;
-
-            // must come last, as it references some _variables
+            // The scheduled tasks - TouchServerTask and InstructionProcessTask - run as .NET Core hosted services.
+            // The former (as well as other hosted services that run outside of an HTTP request context) depends on the application URL
+            // being available (via IRequestAccessor), which can only be retrieved within an HTTP request (unless it's explicitly configured).
+            // Hence we hook up a one-off task on an HTTP request to ensure this is retrieved, which caches the value and makes it available
+            // for the hosted services to use when the HTTP request is not available.
+            _requestAccessor.RouteAttempt += EnsureApplicationUrlOnce;
+
+            // Must come last, as it references some _variables
```

Comment:
I'm in double about whether this is required. The first request was more important in V8, because all the boot of umbraco happed on the first request. Not the server boots without a request.

Reply:
This is the bit I'm in doubt about too, so it does need a careful check.

I can say though we do seem to need this call to `EnsureApplicationUrlOnce`, which fires off some code on the first request.  In particular this calls `_requestAccessor.GetApplicationUrl()`.  Really we don't need to get it here, but by calling this method in the context of an HTTP request, it will find out and cache the application URL for other calls that are made that aren't in an HTTP request - i.e. hosted services.

`TouchServerTask` in particular won't run if it can't get the application URL, so it seems we do need to ensure this is set from the first request.